### PR TITLE
Allow the parameter X to be an empty value

### PR DESCRIPTION
**Pull request by minsunny**
https://nanopay.atlassian.net/browse/CPF-393

- Under foam

Allow the parameter X to be an empty value or whatever.
Since the X is replaced with a real context value later.

### CODE REVIEW

**Comment by deejes** [comment]
Looks good, I will test once and merge. Thanks!

**Comment by deejes** [comment]
Tested, LGTM

**Review by jlhughes — APPROVED** [review]

File: src/foam/nanos/dig/SugarWebAgent.java
Comment (mcarcaso) [inline src/foam/nanos/dig/SugarWebAgent.java:101]: This and the line below are unused. Can we get rid of them?

File: src/foam/nanos/dig/SugarWebAgent.java
Comment (mcarcaso) [inline src/foam/nanos/dig/SugarWebAgent.java:104]: At first, it wasn't clear to me why this condition was being added here and I think it could warrant its own conditional block along with a comment explaining why. How does something like this look?
```
              if ( typeName.equals("foam.core.X") ) {
                // Contexts should not be passed in as arguments so force the use of
                // the current context when a foam.core.X argument is encountered.
                arglist[j] = x;
              } else if ( ! SafetyUtil.isEmpty(p.getParameter(pArray[j].getName())) ) {
                arglist[j] = setParameterType(x, resp, out, p, typeName, pArray[j]);
              } else {
```
If we do this, we can remove the `foam.core.X` case in the `setParameterType` method.

Somewhat unrelated: I was also being a bit thrown off by the name `setParameterType` because it's not really setting a type, it's setting a value based on the type. We might want to consider renaming that.

Comment (minsunny) [reply src/foam/nanos/dig/SugarWebAgent.java:104]: Thank you, let me update when I have a chance asap.

**Comment by deejes** [comment]
@minsunny going to leave it to you resolve mike's issues in another PR.

Thanks.

**Pull request merged**
